plugins/legion-hid2/fu-legion-hid2-device.c

```diff
@@ -0,0 +1,522 @@
+		return TRUE;
+
+	if (!fu_legion_hid2_device_ensure_version(device, error))
+		return FALSE;
+
+	if (!fu_legion_hid2_device_ensure_mcu_id(device, error))
+		return FALSE;
+
+	/* success */
+	return TRUE;
+}
+
+static FuFirmware *
+fu_legion_hid2_device_prepare_firmware(FuDevice *device,
+				       GInputStream *stream,
+				       FuProgress *progress,
+				       FwupdInstallFlags flags,
+				       GError **error)
+{
+	guint32 version;
+	g_autofree gchar *version_str = NULL;
+	g_autoptr(FuFirmware) firmware = fu_legion_hid2_firmware_new();
+
+	/* sanity check */
+	if (!fu_firmware_parse_stream(firmware, stream, 0x0, flags, error))
+		return NULL;
+
+	version = fu_legion_hid2_firmware_get_version(firmware);
+	if (fu_device_get_version_raw(device) > version) {
+		version_str = fu_version_from_uint32(version, FWUPD_VERSION_FORMAT_QUAD);
+		if ((flags & FWUPD_INSTALL_FLAG_FORCE) == 0) {
+			g_set_error(error,
+				    FWUPD_ERROR,
+				    FWUPD_ERROR_NOT_SUPPORTED,
+				    "downgrading from %s to %s is not supported",
+				    fu_device_get_version(device),
+				    version_str);
+			return NULL;
+		}
+		g_warning("firmware %s is a downgrade but is being force installed anyway",
+			  version_str);
+	}
+
+	return g_steal_pointer(&firmware);
+}
+
+static gboolean
+fu_legion_hid2_device_tlv(FuLegionHid2Device *self,
+			  GByteArray *cmd,
+			  GByteArray *result,
+			  GError **error)
+{
+	const guint8 *value;
+	guint8 expected;
+	guint16 tag;
+
+	if (fu_struct_legion_iap_tlv_get_tag(cmd) == FU_LEGION_IAP_HOST_TAG_IAP_UPDATE)
+		expected = FU_LEGION_IAP_ERROR_IAP_CERTIFIED;
+	else
+		expected = FU_LEGION_IAP_ERROR_IAP_OK;
+
+	if (!fu_legion_hid2_device_transfer(self, cmd, result, error))
+		return FALSE;
+
+	tag = fu_struct_legion_iap_tlv_get_tag(result);
+	if (tag != FU_LEGION_IAP_DEVICE_TAG_IAP_ACK) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_WRITE,
+			    "failed to transmit TLV, result: %u",
+			    tag);
+		return FALSE;
+	}
+	value = fu_struct_legion_iap_tlv_get_value(result, NULL);
+	if (value[0] != expected) {
+		g_set_error(error,
+			    FWUPD_ERROR,
+			    FWUPD_ERROR_WRITE,
+			    "failed to transmit TLV, data: %u",
+			    value[0]);
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_legion_hid2_device_unlock_flash(FuLegionHid2Device *self, GError **error)
+{
+	g_autoptr(GByteArray) cmd = fu_struct_legion_iap_tlv_new();
+	g_autoptr(GByteArray) result = fu_struct_legion_iap_tlv_new();
+
+	fu_struct_legion_iap_tlv_set_tag(cmd, FU_LEGION_IAP_HOST_TAG_IAP_UNLOCK);
+
+	if (!fu_legion_hid2_device_tlv(self, cmd, result, error)) {
+		g_prefix_error(error, "failed to unlock: ");
+		return FALSE;
+	}
+
+	return TRUE;
+}
+
+static gboolean
+fu_legion_hid2_device_verify_signature(FuLegionHid2Device *self, GError **error)
+{
+	g_autoptr(GByteArray) cmd = fu_struct_legion_iap_tlv_new();
+	g_autoptr(GByteArray) result = fu_struct_legion_iap_tlv_new();
```

Comment:
do you need the result if it's not being checked?

Reply:
i.e. maybe fu_legion_hid2_device_tlv() should return a GPtrArray rather than bool.